### PR TITLE
all: rework the chain for eth1/2 merging

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR replaces the first version PR #22639 

In order to make the transition from the PoW consensus to PoS consensus, a lot of changes are needed to finish this Docking.
List a few important changes here for easier review:
- Fork choice is abstracted out and not tight with total difficulty anymore.
- Transition upgrade path is handled by `core.Merger` and triggered by the events from consensus layer
- A new consensus engine `beacon` is introduced to follow the post-merge consensus rules
- Block processing is divided into two separate steps after the merge: `InsertBlock` and `SetHeadChain`. All of these actions are controlled by the events from the consensus layer
- Block propagation is changed after the merge. 
- Integrate the Light client mode with catalyst

What's can we do in the next step:
- Enable eth1 style sync for both state and chain after the merge

### CODE REVIEW

File: eth/catalyst/api.go [hunk 1]
```diff
@@ -244,12 +298,11 @@ func decodeTransactions(enc [][]byte) ([]*types.Transaction, error) {
 	return txs, nil
 }
 
-func insertBlockParamsToBlock(params executableData) (*types.Block, error) {
+func InsertBlockParamsToBlock(params ExecutableData) (*types.Block, error) {
```
Comment (mkalinin) [inline eth/catalyst/api.go:301]: Shouldn't this method enforce `block.Hash() == params.BlockHash` check?

Comment (rjl493456442) [reply eth/catalyst/api.go:301]: Fixed!

**Comment by mkalinin** [comment]
Excellent work on the transition process 👍 
I've been playing with it using a [PoC implementation](https://github.com/txrx-research/teku/tree/post-rayonism-merge) in Teku. A couple of screenshots is attached below

Screenshot 2021-06-24 at 15 30 08
Screenshot 2021-06-24 at 15 31 11

**Comment by rjl493456442** [comment]
Close it now. Marius will continue my work in #23607

**Pull request closed without merging**